File: cpp/freelancer-rates/freelancer_rates.cpp
#include <cmath>

// daily_rate calculates the daily rate given an hourly rate
double daily_rate(double hourly_rate) {
    return 8 * hourly_rate;
}

// apply_discount calculates the price after a discount
double apply_discount(double before_discount, double discount) {
    return before_discount - before_discount * discount / 100.0;
}

// monthly_rate calculates the monthly rate, given an hourly rate and a discount
// The returned monthly rate is rounded up to the nearest integer.
int monthly_rate(double hourly_rate, double discount) {
    return int(std::ceil(apply_discount(daily_rate(hourly_rate) * 22, discount)));
}

// days_in_budget calculates the number of workdays given a budget, hourly rate,
// and discount The returned number of days is rounded down (take the floor) to
// the next integer.
int days_in_budget(int budget, double hourly_rate, double discount) {
    return int(std::floor(budget / apply_discount(daily_rate(hourly_rate), discount)));
}

File: elixir/freelancer-rates/lib/freelancer_rates.ex
defmodule FreelancerRates do
  def daily_rate(hourly_rate), do: hourly_rate * 8.0

  defp monthly_rate(hourly_rate), do: daily_rate(hourly_rate) * 22

  def apply_discount(before_discount, discount), do: before_discount - (before_discount * (discount / 100.0))

  def monthly_rate(hourly_rate, discount) do
    hourly_rate |> monthly_rate |> apply_discount(discount) |> ceil
  end

  def days_in_budget(budget, hourly_rate, discount) do
    Float.floor(budget / apply_discount(daily_rate(hourly_rate), discount), 1)
  end
end
